### PR TITLE
dsumaroka/search-books-ability

### PR DESCRIPTION
**Pull request by dsumaroka**

### CODE REVIEW

File: src/main/java/springboot/web/example/controller/BookController.java [hunk 1]
```diff
@@ -32,4 +37,21 @@ public BookDto findById(@PathVariable Long id) {
...
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    public void deleteById(@PathVariable Long id) {
+        bookService.deletedById(id);
+    }
+
+    @GetMapping("/search")
+    List<BookDto> getAllByCriteria(SearchBookParamsDto searchBookParamsDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/springboot/web/example/controller/BookController.java]:
```suggestion
    public List<BookDto> getAllByCriteria(SearchBookParamsDto searchBookParamsDto) {
```

File: src/main/java/springboot/web/example/model/Book.java [hunk 2]
```diff
@@ -8,26 +8,31 @@
...
     private String author;
-    @NonNull
-    @Column(unique = true)
+    @Column(nullable = false, unique = true)
     private String isbn;
-    @NonNull
+    @Column(nullable = false)
     private BigDecimal price;
     private String description;
+    @Column(name = "cover_image")
```
Comment (Elena-Bruyako) [inline src/main/java/springboot/web/example/model/Book.java]:
```suggestion
```

File: src/main/java/springboot/web/example/model/Book.java [hunk 3]
```diff
@@ -8,26 +8,31 @@
...
     private String coverImage;
+    @Column(name = "is_deleted", nullable = false, columnDefinition = "TINYINT(1)")
```
Comment (Elena-Bruyako) [inline src/main/java/springboot/web/example/model/Book.java]:
```suggestion
    @Column(nullable = false, columnDefinition = "TINYINT(1)")
```

File: src/main/java/springboot/web/example/repository/book/SearchBookDtoParser.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.Map;
+import org.springframework.stereotype.Component;
+import springboot.web.example.dto.SearchBookParamsDto;
+
+@Component
+public class SearchBookDtoParser {
+    public Map<String, String[]> parseSearchBookDto(SearchBookParamsDto searchBookParamsDto) {
+        Map<String, String[]> mapOfDtos = new HashMap<>();
+        if (searchBookParamsDto.author() != null && searchBookParamsDto.author().length > 0) {
+            mapOfDtos.put("author", searchBookParamsDto.author());
```
Comment (Elena-Bruyako) [inline src/main/java/springboot/web/example/repository/book/SearchBookDtoParser.java]: create constants for author, title, isbn

File: src/main/java/springboot/web/example/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -34,7 +38,31 @@ public List<BookDto> findAll() {
...
+
+    @Override
+    public void deletedById(Long id) {
+        bookRepository.deleteById(id);
+    }
+
+    @Override
+    public List<BookDto> findAllByCriteria(SearchBookParamsDto searchBookParamsDto) {
+        Specification<Book> bookSpecification = specBuilder.getSpecification(searchBookParamsDto);
+        List<Book> bookListCr = bookRepository.findAll(bookSpecification);
```
Comment (Elena-Bruyako) [inline src/main/java/springboot/web/example/service/impl/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAll(bookSpecification).....
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by dsumaroka** [comment]
Ok

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Никаких
вирусов.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

пн, 7 окт. 2024 г. в 14:54, Olena Bruiako ***@***.***>:

**Pull request merged**